### PR TITLE
fix (faviconfeed): #3834 exponential backoff for tippytop manifest fetch

### LINKED ISSUE #3834: Better handling of failure cases for remote tippy top fetch
Right now, the `FaviconFeed` can get into a state where it tries to fetch on every `SYSTEM_TICK` event. This will happen if there is no cache or the cached data is old and the remote endpoint returns a status code that isn't 200 (ok) or 304 (not modified).

We should probably add some sort of backoff algorithm.

### PR DESCRIPTION
**Pull request by rlr**
@Mardak exponential backoff?

Fixes #3834

### CODE REVIEW

File: system-addon/lib/FaviconFeed.jsm [hunk 1]
```diff
@@ -50,15 +54,20 @@ this.FaviconFeed = class FaviconFeed {
...
       sitesByDomain._etag = etag;
-      this.tippyTopLastUpdated = sitesByDomain._timestamp = Date.now();
+      sitesByDomain._timestamp = Date.now();
       this.cache.set("sites", sitesByDomain);
       this._sitesByDomain = sitesByDomain;
+      this.tippyTopNextUpdate = sitesByDomain._timestamp + TIPPYTOP_UPDATE_TIME;
+      this.numRetries = 0;
+    } else {
+      // Something didn't go quite right. Retry with an exponential backoff.
+      let delay = Math.max(TIPPYTOP_UPDATE_TIME, TIPPYTOP_RETRY_DELAY * Math.pow(2, this.numRetries++));
```
Comment (rlr) [inline system-addon/lib/FaviconFeed.jsm]: can use `const`

File: system-addon/test/unit/lib/FaviconFeed.test.js [hunk 2]
```diff
@@ -163,6 +163,16 @@ describe("FaviconFeed", () => {
       assert.deepEqual(feed._sitesByDomain, data);
       assert.notCalled(feed.cache.set);
     });
+    it("should handle server errors", async () => {
+      feed.loadFromURL = sinon.spy(url => ({data: [], status: 500}));
+      await feed.refresh();
+      assert.equal(1, feed.numRetries);
```
Comment (rlr) [inline system-addon/test/unit/lib/FaviconFeed.test.js]: is there a good way to test that the delay is set properly? Monkeypatch (erm, mock) `Date.now()`?

File: system-addon/lib/FaviconFeed.jsm (hunk 1 above)
Comment (Mardak) [inline system-addon/lib/FaviconFeed.jsm]: I'm thinking you wanted a `Math.min` to cap it at 24 hours of `delay`?

Comment (rlr) [reply system-addon/lib/FaviconFeed.jsm]: lol. yes. #fail

File: system-addon/lib/FaviconFeed.jsm [hunk 3]
```diff
@@ -50,15 +54,20 @@ this.FaviconFeed = class FaviconFeed {
     let {data, etag, status} = await this.loadFromURL(this.prefs.get("tippyTop.service.endpoint"), headers);
     if (status === 304) {
       // Not modified, we are done. No need to dispatch actions or update cache. We'll check again in 24+ hours.
-      this.tippyTopLastUpdated = Date.now();
-      return;
-    }
-    if (status === 200 && data) {
+      this.tippyTopNextUpdate = Date.now() + TIPPYTOP_UPDATE_TIME;
+      this.numRetries = 0;
```
Comment (Mardak) [inline system-addon/lib/FaviconFeed.jsm]: This got me wondering if we should be storing some of this state in the cache? In particular, in the common case of `304`, `_timestamp` isn't updated, so every restart (after 24 hours from first `200`) will trigger a refresh?

File: system-addon/test/unit/lib/FaviconFeed.test.js (hunk 2 above)
Comment (Mardak) [reply system-addon/test/unit/lib/FaviconFeed.test.js]: The `clock = sinon.useFakeTimers()` will have `Date.now()` be `0` and you can advance it however you want. You probably don't need to advance and just check that the "next" time is the expected value, e.g., 0 + 5 minutes = 5 minutes.

Comment (rlr) [reply system-addon/test/unit/lib/FaviconFeed.test.js]: ah, cool. If I had known, I would've caught my math fail 😉

File: system-addon/lib/FaviconFeed.jsm (hunk 3 above)
Comment (rlr) [reply system-addon/lib/FaviconFeed.jsm]: yeah, good point. We should probably store the next update time at least on 304 and 200. Not sure if it's worth writing to cache on every failed retry though?

Comment (Mardak) [reply system-addon/lib/FaviconFeed.jsm]: Ehh.. mm.. I suppose just updating `_timestamp` on `304` should be enough and not bother caching with `numRetries`, i.e., a restart resets the backoff to 5 minutes.

**Comment by rlr** [comment]
@Mardak ^ fixes pushed

File: system-addon/lib/FaviconFeed.jsm [hunk 4]
```diff
@@ -49,16 +53,23 @@ this.FaviconFeed = class FaviconFeed {
...
-      this.tippyTopLastUpdated = sitesByDomain._timestamp = Date.now();
+      sitesByDomain._timestamp = Date.now();
       this.cache.set("sites", sitesByDomain);
       this._sitesByDomain = sitesByDomain;
+      this.tippyTopNextUpdate = sitesByDomain._timestamp + TIPPYTOP_UPDATE_TIME;
+      this.numRetries = 0;
+    } else {
+      // Something didn't go quite right. Retry with an exponential backoff.
+      const delay = Math.min(TIPPYTOP_UPDATE_TIME, TIPPYTOP_RETRY_DELAY * Math.pow(2, this.numRetries++));
+      this.tippyTopNextUpdate = Date.now() + delay;
```
Comment (Mardak) [inline system-addon/lib/FaviconFeed.jsm]: nit: let's avoid the duplicate code:
```js
let failedUpdate = false;
if (status === 200) {
  this._sitesByDomain = this._sitesArrayToObjectByDomain(data);
  this._sitesByDomain._etag = …
} else if (status !== 304) {
  failedUpdate = true;
}
let delay = TIPPYTOP_UPDATE_TIME;
if (failedUpdate) {
  delay = …
} else {
  this._sitesByDomain._timestamp = …
  this.cache.set(…
  this.numRetries = 0;
}
this.tippyTopNextUpdate = Date.now() + delay;
```

**Comment by as-pine-proxy** [comment]
[Treeherder: 2adc00aed6d67c8baad3f9db4d849e04f7e5abfe](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=2adc00aed6d67c8baad3f9db4d849e04f7e5abfe)

**Pull request merged**